Adding notebooks describing how to run Archai on AzureML

luisffranca opened the pull request:
## Description

Adding Tutorial Notebooks showing how to run Archai on AzureML.

## Changes

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update

gugarosa commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,132 @@
+        ml_client: MLClient,
+        custom_env_name: str = "aml-archai",
+        description: str = "Custom environment for Archai",
+        tags: dict = {"archai": "1.0.0"},
```
Personally, I don't like pre-defined lists or dicts. For example:

```
tags: Optional[Dict[str, Any]] = None,
...
tags = tags or {"archai": "1.0.0"}
...
```

But @piero2c and @lovettchris should give their inputs/preferences as well.

luisffranca replied:
Done.

gugarosa commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,132 @@
+    if target_job.status == "Completed":
+        ml_client.jobs.download(name=target_job.name, download_path=Path(download_path), output_name=output_name)
+    else:
+        print(f"Job {target_job.name} is not completed yet")
```
Should we docstring every public method?

luisffranca replied:
Done.

gugarosa commented on docs/advanced_guide/cloud/notebooks/Quickstart/Quickstart.ipynb:
```diff
@@ -0,0 +1,376 @@
+{
```
`docs/advanced_guide/cloud/notebooks/quickstart/` instead of `docs/advanced_guide/cloud/notebooks/Quickstart/`. Just to keep the paths in snake case.

luisffranca replied:
Done.

gugarosa commented on archai/common/notebook_helper.py:
```diff
@@ -0,0 +1,48 @@
+from typing import Union
```
Let's try to keep in every Archai-ownership file:

```
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.`
```

luisffranca replied:
Done.

gugarosa reviewed (changes requested):
Looks good to me! Though I have a single concern:

Since we are having an Amulet-based Text Generation task, would it be useful to be different than https://github.com/microsoft/archai/tree/main/tasks/text_generation?

Different in the sense of the hyperparameters, type of model, some basic things just to avoid having repeated code?

luisffranca commented:
> Looks good to me! Though I have a single concern:
> 
> Since we are having an Amulet-based Text Generation task, would it be useful to be different than https://github.com/microsoft/archai/tree/main/tasks/text_generation?
> 
> Different in the sense of the hyperparameters, type of model, some basic things just to avoid having repeated code?

I agree, repeating the same code wouldn't be a good idea. Our goal was to make this task less expensive so that users don't need to spend a ton of computing resources to be able to execute all steps. This means, however, that the trained models won't be good -- for this purpose the original Text Generation task will be more appropriate. Does that make sense? 

I'll add a note about this on the Notebook.

There are currently a few differences in the code with the original one, such as:

- On `generate_text.py` we added `--output_path`, which will be used to save the generated text to a file. This makes it easier to download the output from AzureML and display it;
- On `search.py` we changed the default number of iterations from 10 to 5;
- On `train.py` we changed the dataset provider and some hyperparams:

`HfHubDatasetProvider(dataset="openwebtext", subset="plain_text")` to `HfHubDatasetProvider(dataset="wikitext", subset="wikitext-2-raw-v1")`

`per_device_train_batch_size` 64 -> 4
`gradient_accumulation_steps=8`
`warmup_steps` 150 -> 5
`max_steps` 30000 -> 1000

Please let me know if this make sense or if we should change something else.

gugarosa commented:
> > Looks good to me! Though I have a single concern:
> > Since we are having an Amulet-based Text Generation task, would it be useful to be different than https://github.com/microsoft/archai/tree/main/tasks/text_generation?
> > Different in the sense of the hyperparameters, type of model, some basic things just to avoid having repeated code?
> 
> I agree, repeating the same code wouldn't be a good idea. Our goal was to make this task less expensive so that users don't need to spend a ton of computing resources to be able to execute all steps. This means, however, that the trained models won't be good -- for this purpose the original Text Generation task will be more appropriate. Does that make sense?
> 
> I'll add a note about this on the Notebook.
> 
> There are currently a few differences in the code with the original one, such as:
> 
> * On `generate_text.py` we added `--output_path`, which will be used to save the generated text to a file. This makes it easier to download the output from AzureML and display it;
> * On `search.py` we changed the default number of iterations from 10 to 5;
> * On `train.py` we changed the dataset provider and some hyperparams:
> 
> `HfHubDatasetProvider(dataset="openwebtext", subset="plain_text")` to `HfHubDatasetProvider(dataset="wikitext", subset="wikitext-2-raw-v1")`
> 
> `per_device_train_batch_size` 64 -> 4 `gradient_accumulation_steps=8` `warmup_steps` 150 -> 5 `max_steps` 30000 -> 1000
> 
> Please let me know if this make sense or if we should change something else.

Sounds perfect!

luisffranca commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,132 @@
+from azure.ai.ml.entities import AmlCompute, Environment, Command, Job
+from azure.identity import DefaultAzureCredential
+
+def get_aml_client_from_file(config_path: Union[str, Path] = "../.azureml/config.json") -> MLClient:
```
Done.

luisffranca commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,132 @@
+def create_compute_cluster(
+        ml_client: MLClient,
+        cpu_compute_name: str,
+        type: str = "amlcompute",
```
Done.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/src/train.py:
```diff
@@ -0,0 +1,89 @@
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description="Trains a Pareto architecture from Transformer-Flex.")
```
Is `Pareto architecture` correct here or should it say `GPT2 architecture` ?

luisffranca replied:
You're right. I changed it to "Trains a GPT2 architecture." and changed the argument's name too to make it consistent.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "ml_client = aml_helper.get_aml_client_from_file(\"../.azureml/config.json\")"
```
I find it safe to add this print statement after this line so user can be reminded they are about to party on the right Azure resources:

```python
print(f'Using workspace: {ml_client.workspace_name} in resource group: {ml_client.resource_group_name}')
```

luisffranca replied:
Done.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+      "source": [
+        "### Create a CPU compute cluster\n",
+        "\n",
+        "We provision a Linux [compute cluster](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-create-attach-compute-cluster?tabs=python) fos the NAS job in this Notebook. See the [full list](https://azure.microsoft.com/en-ca/pricing/details/machine-learning/) on VM sizes and prices."
```
typo "fos the NAS"

luisffranca replied:
Done.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+        }
+      },
+      "source": [
+        "#### Load the search job from a YAML file and run it"
```
a bit more about what to expect here, what it is doing, how long will it run and so on would be nice.

luisffranca replied:
I'll change this Notebook to make it an AML Pipeline, then I'll add this info accordingly.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "#### Stream logs of the job"
```
a bit more detail here would be nice, like "This cell of your notebook will automatically complete when the job is finished".

luisffranca replied:
Done.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "#### Pick an architecture id (archid) from the CSV file to perform full training on"
```
why does this need to be a heading?  I'd remove the '####'

luisffranca replied:
Done.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "#### Load the training job from a YAML file and set the arch_path as its input"
```
why does this need to be a heading?  I'd remove the '####'

luisffranca replied:
Done.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "#### Stream logs of the job"
```
and wait for it to complete

luisffranca replied:
Done.

lovettchris commented on docs/advanced_guide/cloud/notebooks/text_generation/text_generation.ipynb:
```diff
@@ -0,0 +1,469 @@
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "#### Load the generate text job from a YAML file, set the trained model path, and run it"
```
remove '####' and can we figure out this path programatically so the user doesn't have to?

luisffranca replied:
I'll change this Notebook to make it an AML Pipeline, then the user won't have to set up this path.

luisffranca replied:
Done.

luisffranca commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,143 @@
+    Args:
+        config_path (Union[str, Path]): Path to the workspace config file
+
+    Returns:
```
I agree. Once we have an agreement with the team, I'll change the docstrings accordingly.

luisffranca commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,143 @@
+        idle_time_before_scale_down (int, optional): How many seconds will the node running after the job termination. Defaults to 180.
+        tier (str, optional): Dedicated or LowPriority. The latter is cheaper but there is a chance of job termination. Defaults to "Dedicated".
+
+    Returns:
```
Done.

luisffranca commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,143 @@
+        size (str, optional): VM Family of the compute cluster. Defaults to "Standard_D14_v2".
+        min_instances (int, optional): Minimum running nodes when there is no job running. Defaults to 0.
+        max_instances (int, optional): Maximum number of nodes in the cluster. Defaults to 4.
+        idle_time_before_scale_down (int, optional): How many seconds will the node running after the job termination. Defaults to 180.
```
Done.

luisffranca commented on archai/common/azureml_helper.py:
```diff
@@ -0,0 +1,143 @@
+        description: Optional[str] = "Custom environment for Archai",
+        tags: Optional[Dict[str, Any]] = None,
+        conda_file: Optional[str] = "conda.yaml",
+        image: Optional[str] = "mcr.microsoft.com/azureml/openmpi3.1.2-ubuntu18.04:latest",
```
Good point. I haven't found an API to do it, so I removed it as a default argument and changed the function calls to provide it. I also added a note in the Notebooks pointing to the [AzureML Containers](https://github.com/Azure/AzureML-Containers) repo.

luisffranca commented on archai/common/notebook_helper.py:
```diff
@@ -0,0 +1,69 @@
+        pd.DataFrame: Pandas dataframe with the search state
+    """
+    if iteration_num == -1:
+        search_csv_path = max(Path(output_path).glob("search_state_*.csv"), key=lambda x: int(x.stem.split("_")[-1]))
```
Done.

luisffranca commented on archai/common/notebook_helper.py:
```diff
@@ -0,0 +1,69 @@
+        Path: Absolute path to the architecture file
+    """
+    if iteration_num == -1:
+        dir_path = max(Path(downloaded_folder).glob("pareto_models_iter_*"), key=lambda x: int(x.stem.split("_")[-1]))
```
Done.

lovettchris commented:
One more request, can you move the notebooks into an `azure` folder under `cloud`, I'm putting my tutorial here:

> ~\archai\docs\advanced_guide\cloud\azure\notebooks\multi_node_search

luisffranca commented:
> One more request, can you move the notebooks into an `azure` folder under `cloud`, I'm putting my tutorial here:
> 
> > ~\archai\docs\advanced_guide\cloud\azure\notebooks\multi_node_search

Done.
